File: fragment/src/main/java/com/psliusar/layers/LayersFragment.java
package com.psliusar.layers;

import android.app.Activity;
import android.os.Bundle;
import android.view.View;
import android.view.ViewGroup;

import androidx.annotation.IdRes;
import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.fragment.app.Fragment;

public abstract class LayersFragment extends Fragment {

    private static final String SAVED_STATE_LAYERS = "LAYERS.SAVED_STATE_LAYERS";

    private Layers layers;
    private boolean layersStateRestored = false;

    @Override
    public void onCreate(@Nullable Bundle state) {
        super.onCreate(state);
        layersStateRestored = state != null;

        final Activity activity = requireActivity();

        final LayersHost host = new LayersHost() {
            @NonNull
            @Override
            public Layers getLayers() {
                return layers;
            }

            @NonNull
            @Override
            public <T extends View> T getView(int viewId) {
                return LayersFragment.this.getView().findViewById(viewId);
            }

            @NonNull
            @Override
            public ViewGroup getDefaultContainer() {
                return LayersFragment.this.getView().findViewById(getDefaultContainerId());
            }

            @NonNull
            @Override
            public Activity getActivity() {
                return activity;
            }

            @Nullable
            @Override
            public Layer<?> getParentLayer() {
                return null;
            }
        };

        layers = new Layers(host, state != null ? state.getBundle(SAVED_STATE_LAYERS) : null);
    }

    @Override
    public void onStart() {
        super.onStart();
        ensureLayerViews();
    }

    @Override
    public void onSaveInstanceState(@NonNull Bundle outState) {
        super.onSaveInstanceState(outState);
        final Bundle layersState = layers.saveState();
        if (layersState != null) {
            outState.putBundle(SAVED_STATE_LAYERS, layersState);
        }
    }

    @Override
    public void onDestroyView() {
        super.onDestroyView();
        layers.destroy();
    }

    public boolean onBackPressed() {
        if (layers.onBackPressed()) {
            return true;
        }
        if (layers.getStackSize() > 1) {
            layers.pop();
            return true;
        }
        return false;
    }

    @NonNull
    public Layers getLayers() {
        ensureLayerViews();
        return layers;
    }

    @IdRes
    protected abstract int getDefaultContainerId();

    private void ensureLayerViews() {
        if (layersStateRestored) {
            layers.resumeView();
            layersStateRestored = false;
        }
    }
}

File: fragment/src/main/java/com/psliusar/layers/LayersFragment.kt
package com.psliusar.layers

import android.app.Activity
import android.os.Bundle
import android.view.View
import android.view.ViewGroup
import androidx.annotation.CallSuper
import androidx.annotation.IdRes
import androidx.annotation.LayoutRes
import androidx.fragment.app.Fragment

private const val SAVED_STATE_LAYERS = "LAYERS.SAVED_STATE_LAYERS"

/**
 * A [Fragment] implementing [LayersHost]. May be used as an isolated container to host Layers.
 * Subclasses must specify an ID of default ViewGroup that will host the default stack of layers.
 */
abstract class LayersFragment(
    @LayoutRes contentLayoutId: Int = 0
) : Fragment(contentLayoutId), LayersHost {

    private var _layers: Layers? = null
    override val layers: Layers
        get() {
            ensureLayerViews()
            return _layers ?: throw IllegalStateException("Fragment is not created")
        }

    override val defaultContainer: ViewGroup
        get() = requireView().findViewById(defaultContainerId)

    override val activity: Activity
        get() = requireActivity()

    override val parentLayer: Layer? = null

    @get:IdRes
    abstract val defaultContainerId: Int

    private var layersStateRestored = false

    override fun onCreate(state: Bundle?) {
        super.onCreate(state)
        layersStateRestored = state != null
        _layers = Layers(this, state?.getBundle(SAVED_STATE_LAYERS))
    }

    override fun onStart() {
        super.onStart()
        ensureLayerViews()
    }

    override fun onSaveInstanceState(outState: Bundle) {
        super.onSaveInstanceState(outState)
        layers.saveState()?.let {
            outState.putBundle(SAVED_STATE_LAYERS, it)
        }
    }

    override fun onDestroyView() {
        super.onDestroyView()
        layers.destroy()
    }

    /**
     * A useful method to pop the stack of layers.
     *
     * @return true if the back press was handled. False if the stack is empty.
     */
    @CallSuper
    open fun onBackPressed(): Boolean =
        _layers?.onBackPressed() == true || _layers?.pop<Layer>() != null

    override fun <T : View> getView(viewId: Int): T = requireView().findViewById(viewId)

    private fun ensureLayerViews() {
        if (layersStateRestored) {
            _layers?.resumeView()
            layersStateRestored = false
        }
    }
}
